feat(patterns): add Error documentation (#451)
zendeskgarden/website: src/pages/patterns/errors.mdx

---
title: Error handling
description: >
  Error messages let users know that a problem has occurred and what to do about it.
---

Examples of errors include:

- The system has failed to load or perform an action
- The user has made an incorrect action or input

Error messages are best surfaced using these components:

- Global Alerts
- Alerts
- Notifications

> - Validation messages (in Forms)

Reviewer: Would it be helpful to link to these components?

## Global Alerts

> Use the Global Alert to communicate system level errors that interrupt a user’s task.

Reviewer: ```suggestion
Use the Global Alert to communicate system level errors that interrupt a users task.
```
I don't think contraction was intended here.

Reviewer 2: I think the apostrophe needs to remain here. https://forum.wordreference.com/threads/user%E2%80%99s-users-or-users.1474044/#post-7448965

Reviewer: Good call @jzempel

### Global alert anatomy

- Always include a title to explain what has went wrong and how to fix it
- Provide a clear action to address the issue

![Global Alert anatomy](figma:patterns-error-global-alert-anatomy.png)

> 1. Title
> 2. Message
> 3. Button

Reviewer 2: ```suggestion
1. Title
1. Message
1. Button
```

[nit] the suggested markdown follows Garden's preference for ordered lists. Consider applying throughout.

<BestPractice>
  <Do
    imageIsFreeForm
    imageSource={props.data.FormsValidateDo.childFile.childImageSharp.gatsbyImageData}
  >
    <ul>
      <li>Show error validation after the user has tried submitting the form.</li>
    </ul>
  </Do>
  <Dont
    imageIsFreeForm
    imageSource={props.data.FormsValidateDont.childFile.childImageSharp.gatsbyImageData}
  >
    <ul>
      <li>Do not use live validation.</li>
    </ul>
  </Dont>
</BestPractice>

#### Keep data the user has entered

Keep incorrect values that users have entered to help them avoid making the same mistake twice.

<BestPractice>
  <Do
    imageIsFreeForm
    imageSource={props.data.FormsKeepDataDo.childFile.childImageSharp.gatsbyImageData}
  >
    <ul>
      <li>Keep invalid data to let users understand what to correct.</li>
    </ul>
  </Do>
  <Dont
    imageIsFreeForm
    imageSource={props.data.FormsKeepDataDont.childFile.childImageSharp.gatsbyImageData}
  >
    <ul>
      <li>Do not clear invalid data.</li>
    </ul>
  </Dont>
</BestPractice>

## Accessibility

Error messages need to be perceivable and understandable for people of all abilities.
An accessible error message should exhibit these qualities.

> - Written in text. Icons and color can be used, but not alone

Reviewer 2: ```suggestion
- Written in text. Icons and color can be used, but not alone.
```

[nit] per https://garden.zendesk.com/content/punctuation#period

- Close to the element that triggered the error
- Informative, helping the user resolve the issue
- The message should refer to the failed element in code

## Additional resources

[https://www.w3.org/WAI/standards-guidelines/act/rules/36b590/proposed/](https://www.w3.org/WAI/standards-guidelines/act/rules/36b590/proposed/)
[https://www.w3.org/WAI/WCAG21/Understanding/error-identification.html](https://www.w3.org/WAI/WCAG21/Understanding/error-identification.html)
[https://www.w3.org/WAI/WCAG21/Understanding/error-suggestion.html](https://www.w3.org/WAI/WCAG21/Understanding/error-suggestion.html)

---

> <h2>Feedback</h2>

Reviewer 2: ```suggestion
## Feedback
```

---
title: Error handling
description: >
  Error messages let users know that a problem has occurred and what to do about it.
---

<!-- markdownlint-disable no-duplicate-header -->

Examples of errors include:

- The system has failed to load or perform an action
- The user has made an incorrect action or input

Error messages are best surfaced using these components:

- [Global Alerts](/components/global-alert)
- [Alerts](/components/alerts)
- [Notifications](/components/notifications)

> - [Validation messages](/components/forms/input#validation)

Reviewer 2: ```suggestion
- [Validation messages](/components/input#validation)
```
